src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs

```diff
@@ -15,7 +15,9 @@ public abstract partial class PipeReader
         /// <summary>Attempts to synchronously read data from the <see cref="System.IO.Pipelines.PipeReader" />.</summary>
         /// <param name="result">When this method returns <see langword="true" />, this value is set to a <see cref="System.IO.Pipelines.ReadResult" /> instance that represents the result of the read call; otherwise, this value is set to <see langword="default" />.</param>
         /// <returns><see langword="true" /> if data was available, or if the call was canceled or the writer was completed; otherwise, <see langword="false" />.</returns>
-        /// <remarks>If the pipe returns <see langword="false" />, there is no need to call <see cref="System.IO.Pipelines.PipeReader.AdvanceTo(System.SequencePosition,System.SequencePosition)" />.</remarks>
+        /// <remarks>If the pipe returns <see langword="false" />, there is no need to call <see cref="System.IO.Pipelines.PipeReader.AdvanceTo(System.SequencePosition,System.SequencePosition)" />.
+        /// If the current <see cref="System.IO.Pipelines.PipeReader" /> implementation is a <see cref="System.IO.Pipelines.StreamPipeReader" />,
+        /// <see cref="System.IO.Pipelines.PipeReader.ReadAsync(System.Threading.CancellationToken)" /> must be called first in order to asynchronously retrieve data from the <see cref="System.IO.Pipelines.PipeReader" />.</remarks>
```

Comment:
PipeReader is an abstraction that can have many implementations. The documentation should not mention just one of them.
Would be better to describe what it does (reads unconsumed bytes of the last ReadAsync?) 
and the proper usage which should work with any conforming implementation.

Reply:
Thanks, I agree that it's somewhat awkward to mention a specific implementation here. I'll update this.

Reply:
@omariom The more I think this through, the more I think that the current approach (or one like it) of being specific to StreamPipeReader is the correct approach. I'm happy to put this documentation in a different place, though, if this method doc is not the best place.

Here's my reasoning:

1. I believe the intention of `PipeReader.TryRead` is support reading synchronously (i.e. not in conjunction with `ReadAsync`). This is according to the current method `<summary>` documentation, as well as implementations [here](https://github.com/mgravell/Pipelines.Sockets.Unofficial/blob/19b141f30fe2485dd79993be0ac87a9fc6b34b74/src/Pipelines.Sockets.Unofficial/MemoryMappedPipeReader.cs#L217) and [here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs#L74).
2. I think StreamPipeReader does not have the same semantics, but it's a very popular implementation. See https://stackoverflow.com/questions/64820259 for some of the confusion I'm trying to clear up.

I don't think we should document StreamPipeReader's current semantics as the intended semantics of the base class. Any thoughts on the above? Thanks.

Reply:
I was tempted to ask for this to be put on PipeReader.Create, but I think it's better here where it is because this is really a gotcha for user's PipeReader.TryRead who very well might not know where the PipeReader came from. ASP.NET Core's HttpContext.Request.BodyReader is sometimes backed by StreamPipeReader and sometimes backed by DefaultPipeReader for example.

That said, StreamPipeReader is an internal type, so we shouldn't reference it directly.

How about something like this instead with the appropriate `<see ...` markup?

> The PipeReader implementation returned by `PipeReader.Create(Stream, StreamPipeReaderOptions?)` will not read new data from the backing Stream when TryRead is called. PipeReader.ReadAsync must be called to read new data from the backing Stream. Any unconsumed data from a previous read should always be available from `TryRead`.

Also, can we make this a "note" or something so it stands apart from the rest of the docs for the method which are more about general usage?

Reply:
Thanks @halter73, that's a good way to avoid mentioning the internal type. I've pushed a change that's basically your wording, with the final sentence tweaked a bit to clearly differentiate between the async/sync read calls. Thanks!